### ISSUE TITLE
try to learn a simple function (regression)

### ISSUE THREAD

**Issue by ucacaxm** [Microsoft/CNTK#876]
hi, 

I'm still confused by how to use cntk to learn a function (regression)
I moved to a simple one: input is [0,10] and ouput is [0,10] with cosine :
		x = frand(0.f, 10.f);
		y = 5.f * (1 + cos(2.0*M_PI*x/10.f));

I have 100000 data.
I read the post : https://github.com/Microsoft/CNTK/issues/699
which seems to have the same kind of difficulties ... 

My Squared error is always around 30/40 and never decrease. 
I tried with a learningRate of 0.1, 1 or 10; also with miniBatch=10, 128 or 1000
My Cntk files and data are here [cntk_cos.zip](https://github.com/Microsoft/CNTK/files/497557/cntk_cos.zip)

IK_TrainXY = [
    action = "train"

    # network description
    BrainScriptNetworkBuilder=[

		DNNReluLayer (inDim, outDim, x, wScale) = [
			W = LearnableParameter(outDim, inDim, init = "uniform", initValueScale=wScale ) 
			b = LearnableParameter(outDim, 1, init = "uniform", initValueScale=wScale) 
			t = W * x + b
			y = RectifiedLinear(t)
		].y

        # sample and label dimensions
        featureDim = 1
        labelDim = 1
	hiddenDim = 200

        features = Input (featureDim)
        labels   = Input (labelDim)
		
	h1 = DNNReluLayer ( featureDim,   hiddenDim, features, 1)
	z  = DNNReluLayer ( hiddenDim, labelDim,  h1, 1)
				
				
	cri = SquareError(labels,z)
        eval = SquareError (labels, z)
	#eval = ErrorPrediction(labels, z)
		
        # root nodes
        featureNodes    = (features)
        labelNodes      = (labels)
        criterionNodes  = (cri)
        evaluationNodes = (eval)
        outputNodes     = (z)

    ]

	
    SGD = [
		modelPath = "$modelPath$"
        	epochSize = 60000
	        minibatchSize = 10
        	learningRatesPerSample = 0.5
	        maxEpochs = 3000
        	momentumPerMB = 0           
		
		#numMBsToShowResult = 100  # show intermediate objective values every 100 minibatches
		#firstMBsToShowResult = 10  # and for the first 10 minibatches
	]
	
	

    # Parameter values for the reader		
	reader = [
		readerType = "CNTKTextFormatReader"
        file = "$DataDir$/cosAnalyticData.txt"
		input = [
			features = [
				dim = 1
				format = "dense"
			]
			labels = [
				dim = 1
				#labelType = "Regression"
				format = "dense"
			]
		]
	]

]

**Comment by mfuntowicz** [comment]
Hi ucacaxm, 

Some general comments first:

- You're trying to train a regression model with a ReLU activation at the end. With regression we generally use a linear activation. By using ReLU on the last layer, if your model output a value < 0, the ReLU will output 0, which will backpropagate a gradient of 0, which will not improve anything.
- You can add a momentum to guide the gradient according to some previous update.
- Finally 200 neurons might be too much for this kind of task, same for 3000 epoch.

I modified your cos.cntk file, and annotated the modifications I made. 
Your model is now able to get a SquaredError of 0.10 for a 3% error on average.

Do not hesitate to come back on this thread if you have any question :)

Morgan

```

RootDir = ".."

ConfigDir = "$RootDir$/Config"
DataDir = "$RootDir$/Data"
OutputDir = "$RootDir$/Output"
ModelDir = "$OutputDir$/Models"

# deviceId=-1 for CPU, >=0 for GPU devices, "auto" chooses the best GPU, or CPU if no usable GPU is available
deviceId = -1

command = IK_TrainXY


precision = "float"
traceLevel = 1
modelPath = "$ModelDir$/cos.dnn"
outputNodeNames = ScaledLogLikelihood

#######################################
#  TRAINING CONFIG                    #
#######################################

IK_TrainXY = {
    action = "train"

    # network description
    BrainScriptNetworkBuilder= {
        # sample and label dimensions
        featureDim = 1
        labelDim = 1
		hiddenDim = 10 #Less neurons are easier and faster to optimize, reduce overfitting

        features = Input (featureDim)
        labels   = Input (labelDim)

		# I use the Sequential declaration, you need CNTK >= 1.7
		CosModel = Sequential(
			FeatureMVNLayer {} : 	# I also added a normalization layer, improving the optimization
			DenseLayer { hiddenDim, init='uniform', activation=ReLU } :
			LinearLayer{ labelDim, init='uniform' }
		)

		# I call my pipeline with the data you provided, stored in "features" variable
		z = CosModel(features)

		# You used Squared as training and eval criteron, so no need to compute twice, let's reuse				
		cri   = SquareError (labels, z)
		
        # root nodes
        featureNodes    = (features)
        labelNodes      = (labels)
        criterionNodes  = (cri)
        evaluationNodes = (Sqrt(cri)) 			# The error I use is the square root of the Squared error, which give you residual between pred and real values
        outputNodes     = (z)
    }

	
    SGD = {
		modelPath = "$modelPath$"
        epochSize = 10000                    	# I reduced the epochsize just to make tests faster
        minibatchSize = 128
        learningRatesPerSample = 0.000125 		# I reduced the learningRatesPerSample which was really big
        maxEpochs = 20  						# I reduced the number of iterations done by the optimizer. It doesn't learn anything more after 20 iters. 
        momentumPerMB = 0.9 					# I added a momentum, which give better optimization results
		
		numMBsToShowResult = 16  				# show intermediate objective values every 16 minibatches
		firstMBsToShowResult = 10  				# and for the first 10 minibatches
	}
	
	


    # Parameter values for the reader		
	reader = {
		readerType = "CNTKTextFormatReader"
        file = "$DataDir$/cosAnalyticData.txt"
		input = [
			features = [
				dim = 1
				format = "dense"
			]
			labels = [
				dim = 1
				#labelType = "Regression"
				format = "dense"
			]
		]
	}

}
```

**Comment by ucacaxm** [comment]
Thank very much for your feat back, I really appreciate, it is exactly what I needed ...
I definitely need to practice NN
Thanks a lot

**Comment by mfuntowicz** [comment]
Glad to see that it helped :).

Im closing this issue, dont hesitate to open it again if needed.

Morgan
